FreeBSD9/lib/libc/sys/rtprio.2
.\" Copyright (c) 1994, Henrik Vestergaard Draboel
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\" 	This product includes software developed by Henrik Vestergaard Draboel.
.\" 4. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: release/9.0.0/lib/libc/sys/rtprio.2 140505 2005-01-20 09:17:07Z ru $
.\"
.Dd July 23, 1994
.Dt RTPRIO 2
.Os
.Sh NAME
.Nm rtprio
.Nd examine or modify a process realtime or idle priority
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In sys/types.h
.In sys/rtprio.h
.Ft int
.Fn rtprio "int function" "pid_t pid" "struct rtprio *rtp"
.Sh DESCRIPTION
The
.Fn rtprio
system call
is used to lookup or change the realtime or idle priority of a process.
.Pp
The
.Fa function
argument
specifies the operation to be performed.
RTP_LOOKUP to lookup the current priority,
and RTP_SET to set the priority.
The
.Fa pid
argument
specifies the process to be used, 0 for the current process.
.Pp
The
.Fa *rtp
argument
is a pointer to a struct rtprio which is used to specify the priority and priority type.
This structure has the following form:
.Bd -literal
struct rtprio {
	u_short	type;
	u_short prio;
};
.Ed
.Pp
The value of the
.Va type
field may be RTP_PRIO_REALTIME for realtime priorities,
RTP_PRIO_NORMAL for normal priorities, and RTP_PRIO_IDLE for idle priorities.
The priority specified by the
.Va prio
field ranges between 0 and
.Dv RTP_PRIO_MAX (usually 31) .
0 is the highest possible priority.
.Pp
Realtime and idle priority is inherited through fork() and exec().
.Pp
A realtime process can only be preempted by a process of equal or
higher priority, or by an interrupt; idle priority processes will run only
when no other real/normal priority process is runnable.
Higher real/idle priority processes
preempt lower real/idle priority processes.
Processes of equal real/idle priority are run round-robin.
.Sh RETURN VALUES
.Rv -std rtprio
.Sh ERRORS
The
.Fn rtprio
system call
will fail if
.Bl -tag -width Er
.It Bq Er EINVAL
The specified
.Fa prio
was out of range.
.It Bq Er EPERM
The calling process is not allowed to set the realtime priority.
Only
root is allowed to change the realtime priority of any process, and non-root
may only change the idle priority of the current process.
.It Bq Er ESRCH
The specified process was not found.
.El
.Sh SEE ALSO
.Xr nice 1 ,
.Xr ps 1 ,
.Xr rtprio 1 ,
.Xr setpriority 2 ,
.Xr nice 3 ,
.Xr renice 8
.Sh AUTHORS
.An -nosplit
The original author was
.An Henrik Vestergaard Draboel Aq hvd@terry.ping.dk .
This implementation in
.Fx
was substantially rewritten by
.An David Greenman .


FreeBSD9/usr.sbin/rtprio/rtprio.c
/*
 * Copyright (c) 1994 David Greenman
 * Copyright (c) 1994 Henrik Vestergaard Draboel (hvd@terry.ping.dk)
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by Henrik Vestergaard Draboel.
 *	This product includes software developed by David Greenman.
 * 4. Neither the names of the authors nor the names of contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__FBSDID("$FreeBSD: release/9.0.0/usr.sbin/rtprio/rtprio.c 216967 2011-01-04 17:27:17Z kib $");

#include <sys/param.h>
#include <sys/rtprio.h>

#include <ctype.h>
#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

static int parseint(const char *, const char *);
static void usage(void);

int
main(int argc, char *argv[])
{
	struct rtprio rtp;
	char *p;
	pid_t proc;

	/* find basename */
	if ((p = rindex(argv[0], '/')) == NULL)
		p = argv[0];
	else
		++p;
	proc = 0;

	if (!strcmp(p, "rtprio"))
		rtp.type = RTP_PRIO_REALTIME;
	else if (!strcmp(p, "idprio"))
		rtp.type = RTP_PRIO_IDLE;

	switch (argc) {
	case 2:
		proc = parseint(argv[1], "pid");
		proc = abs(proc);
		/* FALLTHROUGH */
	case 1:
		if (rtprio(RTP_LOOKUP, proc, &rtp) != 0)
			err(1, "RTP_LOOKUP");
		printf("%s: ", p);
		switch (rtp.type) {
		case RTP_PRIO_REALTIME:
		case RTP_PRIO_FIFO:
			printf("realtime priority %d\n", rtp.prio);
			break;
		case RTP_PRIO_NORMAL:
			printf("normal priority\n");
			break;
		case RTP_PRIO_IDLE:
			printf("idle priority %d\n", rtp.prio);
			break;
		default:
			printf("invalid priority type %d\n", rtp.type);
			break;
		}
		exit(0);
	default:
		if (argv[1][0] == '-' || isdigit(argv[1][0])) {
			if (argv[1][0] == '-') {
				if (strcmp(argv[1], "-t") == 0) {
					rtp.type = RTP_PRIO_NORMAL;
					rtp.prio = 0;
				} else {
					usage();
					break;
				}
			} else
				rtp.prio = parseint(argv[1], "priority");
		} else {
			usage();
			break;
		}

		if (argv[2][0] == '-')
			proc = parseint(argv[2] + 1, "pid");
		if (rtprio(RTP_SET, proc, &rtp) != 0)
			err(1, "RTP_SET");

		if (proc == 0) {
			execvp(argv[2], &argv[2]);
			err(1, "%s", argv[2]);
		}
		exit(0);
	}
	exit(1);
}

static int
parseint(const char *str, const char *errname)
{
	char *endp;
	long res;

	errno = 0;
	res = strtol(str, &endp, 10);
	if (errno != 0 || endp == str || *endp != '\0')
		errx(1, "%s must be a number", errname);
	if (res >= INT_MAX)
		errx(1, "Integer overflow parsing %s", errname);
	return (res);
}

static void
usage(void)
{

	(void) fprintf(stderr, "%s\n%s\n%s\n%s\n%s\n%s\n",
		"usage: [id|rt]prio",
		"       [id|rt]prio [-]pid",
		"       [id|rt]prio priority command [args]",
		"       [id|rt]prio priority -pid",
		"       [id|rt]prio -t command [args]",
		"       [id|rt]prio -t -pid");
	exit(1);
}
